Don't create empty holding for MPT issuer. Enforce MPToken creation by non-issuer invariant.

gregtatcam opened the pull request:
## High Level Overview of Change

This fixes [RIPD-3462.](https://ripplelabs.atlassian.net/issues/?filter=-1&selectedIssue=RIPD-3462)

### Context of Change

Current addEmptyHolding() creates MPToken for an issuer. This is incorrect since an issuer can't create MPToken. Updated version of addEmptyHolding() returns tesSUCCESS if an account is the issuer.

Updated ValidMPTIssuance class in InvariantCheck enforces MPToken creation by non-issuer invariant.

gregtatcam committed:
Don't create empty holding for MPT issuer.
Enhance MPToken creation invariant by non-issuer.

Bronek commented:
Please see #5875

gregtatcam committed:
Allow addEmptyHolding() to succeed for IOU issuer, but not create the trustline.

a1q123456 commented on src/xrpld/app/tx/detail/InvariantCheck.cpp:
```diff
@@ -1493,6 +1498,15 @@ ValidMPTIssuance::finalize(
+            JLOG(j.fatal())
+                << "Invariant failed: MPT authorize submitted by issuer "
+                   "succeeded but created mptokens";
+            return false;
```
QQ: What's the difference between failing here and failing in `ValidMPTIssuance::visitEntry`

shawnxie999 replied:
i don't think you can fail in visitEntry, it's a void function

a1q123456 replied:
Ah yeah, thanks, Shawn!

shawnxie999 commented on src/libxrpl/ledger/View.cpp:
```diff
@@ -1282,6 +1282,8 @@ addEmptyHolding(
         return tefINTERNAL;  // LCOV_EXCL_LINE
     if (view.peek(keylet::mptoken(mptID, accountID)))
         return tecDUPLICATE;
+    if (accountID == mptIssue.getIssuer())
+        return tesSUCCESS;
```
is `tesSuccess` really the correct code to return here? From my intuition, an issuer should not be allowed to add an empty holding, therefore an error code should be returned instead - and the caller of `addEmptyHolding` should be responsible for checking that the accountID is not the issuer.

a1q123456 replied:
I think this is the reason: https://github.com/XRPLF/rippled/pull/5875/files#r2426305173

shawnxie999 replied:
although it's clean to handle this edge case in the function, i still think it might be something that's worth being explicitly handled by the caller for safety practice, as there could be downstream bugs if future callers aren't aware that this edge case is silently returned as tesSuccess. But again, we do have sufficient unit tests and invariants for so the likelihood of something like this going wrong is pretty low.

ximinez replied:
`temDST_IS_SRC` is the result returned by a `TrustSet` transaction if the account is the issuer. ~I think the situation is analogous enough that it would be appropriate here, too.~

This change does _not_ need to be amendment gated, because `addEmptyHolding` is only used with SAV and LP, which are already amendment gated.

shawnxie999 commented on src/xrpld/app/tx/detail/InvariantCheck.cpp:
```diff
@@ -1493,6 +1498,15 @@ ValidMPTIssuance::finalize(
+            JLOG(j.fatal())
+                << "Invariant failed: MPT authorize submitted by issuer "
+                   "succeeded but created mptokens";
+            return false;
```
i believe this needs to be amendment gated

ximinez replied:
> i believe this needs to be amendment gated

Agreed. I would have the gate check for either `SingleAssetVault` or `LendingProtocol` to be enabled.

shawnxie999 reviewed (changes requested):
The invariant needs to be amendment gated (by SingleAssetVault?). Also, I don’t think we should silently return `tesSUCCESS` in `addEmptyHolding` when the account is the issuer - this could potentially lead to misuse in the future, where future callers of this function might overlook this edge case. I feel this condition should be explicitly handled by the caller instead. That said, this is subjective and not a blocker.

ximinez commented on src/libxrpl/ledger/View.cpp:
```diff
@@ -1226,8 +1226,8 @@ addEmptyHolding(
 {
-    // Every account can hold XRP.
-    if (issue.native())
+    // Every account can hold XRP. An issuer can issue directly.
+    if (issue.native() || accountID == issue.getIssuer())
         return tesSUCCESS;
```
~I think `temDST_IS_SRC` should be returned here, too.~

This change does _not_ need to be amendment gated, because `addEmptyHolding` is only used with SAV and LP, which are already amendment gated.

ximinez commented on src/test/app/Loan_test.cpp:
```diff
@@ -2709,10 +2703,51 @@ class Loan_test : public beast::unit_test::suite
+        auto const broker = createVaultAndBroker(env, asset, lender, 200);
+        auto const loanSetFee = fee(env.current()->fees().base * 2);
+        // Create Loan
+        env(set(borrower, broker.brokerID, 200),
+            sig(sfCounterpartySignature, lender),
+            loanSetFee);
+        env.close();
+        // Issuer should not create MPToken
+        BEAST_EXPECT(!env.le(keylet::mptoken(asset.issuanceID(), issuer)));
+        // Issuer "borrowed" 200, OutstandingAmount decreased by 200
+        BEAST_EXPECT(env.balance(issuer, asset) == asset(-issuerBalance + 200));
+        // Pay Loan
+        auto const loanKeylet = keylet::loan(broker.brokerID, 1);
+        env(pay(borrower, loanKeylet.key, asset(200)));
+        env.close();
+        // Issuer "re-payed" 200, OutstandingAmount increased by 200
+        BEAST_EXPECT(env.balance(issuer, asset) == asset(-issuerBalance));
```
Wait. You mean that by just avoiding creating the holding, everything else "just works"!?!?

ximinez commented:
You know, on further thought. I think `tesSUCCESS` is the way to go.

My thinking is that if the caller needs to do something different to handle the case where the issuer is the relevant account, then `addEmptyHolding` should fail, and the callers should check the account, and do the alternate thing if it's the issuer.

But that's not the case. As your unit test demonstrates, nothing else needs to be done. Therefore, it is a complete solution to have `addEmptyHolding` return `tesSUCCESS` if the account is the issuer. In fact, that's better than checking in the caller, because then you don't have to worry about forgetting.

The one caveat is that this also has to hold true for `VaultCreate`. I suspect it does, because it's doing the same basic operations (and I stole a bunch of code from from it to start implementing these transactions). @Bronek Do you have any unit tests checking that scenario?

ximinez committed:
Merge branch 'ximinez/lending-XLS-66' into gregtatcam/bugs/ripd-3462

ximinez committed:
Merge branch 'ximinez/lending-XLS-66' into gregtatcam/bugs/ripd-3462

Bronek commented:
> You know, on further thought. I think `tesSUCCESS` is the way to go.
> 
> My thinking is that if the caller needs to do something different to handle the case where the issuer is the relevant account, then `addEmptyHolding` should fail, and the callers should check the account, and do the alternate thing if it's the issuer.
> 
> But that's not the case. As your unit test demonstrates, nothing else needs to be done. Therefore, it is a complete solution to have `addEmptyHolding` return `tesSUCCESS` if the account is the issuer. In fact, that's better than checking in the caller, because then you don't have to worry about forgetting.
> 
> The one caveat is that this also has to hold true for `VaultCreate`. I suspect it does, because it's doing the same basic operations (and I stole a bunch of code from from it to start implementing these transactions). @Bronek Do you have any unit tests checking that scenario?

In `VaultCreate` we only call `addEmptyHolding` for the pseudo-account, which by definition cannot be an issuer of the assets stored in vault (logic: `canAddHolding` called inside `preclaim`, before we even created the pseudo-account). 

However this does impact `VaultWithdraw` where we call `addEmptyHolding` to ensure that there is somewhere to store the balance of the withdrawn assets. This is happening in https://github.com/XRPLF/rippled/blob/5e33ca56fd51b02a5490e8e156e2de65937cf88e/src/xrpld/app/tx/detail/VaultWithdraw.cpp#L296-L304

... and the relevant unit tests are in https://github.com/XRPLF/rippled/blob/5e33ca56fd51b02a5490e8e156e2de65937cf88e/src/test/app/Vault_test.cpp#L420-L439

I think with this change, the above displayed part of `VaultWithdraw` should be simplified.

ximinez commented:
> I think with this change, the above displayed part of VaultWithdraw should be simplified.

Do you want to tack a commit making that change onto this PR?

ximinez committed:
Merge branch 'ximinez/lending-XLS-66' into gregtatcam/bugs/ripd-3462

ximinez committed:
Merge remote-tracking branch 'mywork/ximinez/lending-XLS-66' into upstream--gregtatcam/bugs/ripd-3462

* mywork/ximinez/lending-XLS-66:
  Extend LoanBroaker and Loan unit-tests. (5863)

dangell7 commented:
> You know, on further thought. I think `tesSUCCESS` is the way to go.
> 
> My thinking is that if the caller needs to do something different to handle the case where the issuer is the relevant account, then `addEmptyHolding` should fail, and the callers should check the account, and do the alternate thing if it's the issuer.
> 
> But that's not the case. As your unit test demonstrates, nothing else needs to be done. Therefore, it is a complete solution to have `addEmptyHolding` return `tesSUCCESS` if the account is the issuer. In fact, that's better than checking in the caller, because then you don't have to worry about forgetting.
> 
> The one caveat is that this also has to hold true for `VaultCreate`. I suspect it does, because it's doing the same basic operations (and I stole a bunch of code from from it to start implementing these transactions). @Bronek Do you have any unit tests checking that scenario?

See I actually go the other way a bit. If this was only supposed to be used in one place maybe return tesSUCCESS but if its agnostic then it probably should error I feel. However there is a GlobalFreeze check in there that makes it not agnostic because when locking tokens PayChan/Escrow/Offer/Subscriptions we handle GlobalFreeze differently based on who's receiving the funds.

I built this function the other day. I would really love to see us move towards this type of thing.

https://github.com/XRPLF/rippled/blob/0039fdb83fb379ee00c2e84e80fa5cd39e52ba7f/src/libxrpl/ledger/View.cpp#L2960

idk

ximinez commented:
> See I actually go the other way a bit. If this was only supposed to be used in one place maybe return tesSUCCESS but if its agnostic then it probably should error I feel.

Consider that if you pass an `Asset` representing XRP to `addEmptyHolding`, it succeeds. Because there's nothing to do. The same applies if the account is the issuer.

ximinez committed:
Gate the new MPT invariant on amendments

- Checks that MPToken is not created by MPTokenIssuance issuer. Fails if
  SAV or LP amendments are enabled.

ximinez committed:
Merge remote-tracking branch 'mywork/ximinez/lending-XLS-66' into gregtatcam/bugs/ripd-3462

* mywork/ximinez/lending-XLS-66:
  Fix broken Vault unit tests introduced in 6adb2ec
  Review feedback from @Tapanito
  Review feedbak from @tequdev, plus helpers

ximinez committed:
Merge remote-tracking branch 'mywork/ximinez/lending-XLS-66' into gregtatcam/bugs/ripd-3462

* mywork/ximinez/lending-XLS-66:
  Fix more problems introduced by 6adb2ec
  test: Count crashed test suites
